### PR TITLE
net/ip/dhcpv4: Set source IP address for DHCP renewal requests

### PR DESCRIPTION
**Pull request by pfl**
The source address in unicast DHCPv4 Request packets was found out
to be all zeros address 0.0.0.0. This address is only acceptable if
the destination is a multicast one, where the host in question is
acquiring a DHCP address lease. This is true for the DHCP Discover
and the initial DHCP Request message from the client towards the
server. As subsequent DHCP Request renewal messages are sent as
unicast to the server, the server will drop such packets.

Fix this issue by explicitely specifying what source IP address is
to be used, if none is specified, the all zeros address 0.0.0.0 is
used in multicast addresses. The source address in the other
unicast cases is identical to the 'ciaddr' in the DHCP message.

Signed-off-by: Patrik Flykt <patrik.flykt@intel.com>

### CODE REVIEW

File: subsys/net/ip/dhcpv4.c [hunk 1]
```diff
@@ -150,15 +150,22 @@ static inline bool dhcpv4_add_sname(struct net_pkt *pkt)
...
 {
 	NET_PKT_DATA_ACCESS_DEFINE(dhcp_access, struct dhcp_msg);
-	const struct in_addr src = INADDR_ANY_INIT;
+	const struct in_addr any = INADDR_ANY_INIT;
+	const struct in_addr *addr;
 	size_t size = DHCPV4_MESSAGE_SIZE;
 	struct net_pkt *pkt;
 	struct dhcp_msg *msg;
 
+	if (src == NULL)
```
Comment (tbursztyka) [inline subsys/net/ip/dhcpv4.c]: Missing {}

File: subsys/net/ip/dhcpv4.c [hunk 2]
```diff
@@ -150,15 +150,22 @@ static inline bool dhcpv4_add_sname(struct net_pkt *pkt)
...
+		addr = &any;
+	else
```
Comment (tbursztyka) [inline subsys/net/ip/dhcpv4.c]: ditto

File: subsys/net/ip/dhcpv4.c [hunk 3]
```diff
@@ -150,15 +150,22 @@ static inline bool dhcpv4_add_sname(struct net_pkt *pkt)
 /* Create DHCPv4 message and add options as per message type */
 static struct net_pkt *dhcpv4_create_message(struct net_if *iface, u8_t type,
 					     const struct in_addr *ciaddr,
+					     const struct in_addr *src,
```
Comment (tbursztyka) [inline subsys/net/ip/dhcpv4.c]: maybe src_addr would be a nicer name?

File: subsys/net/ip/dhcpv4.c [hunk 4]
```diff
@@ -245,6 +252,7 @@ static struct net_pkt *dhcpv4_create_message(struct net_if *iface, u8_t type,
 /* Prepare DHCPv4 Message request and send it to peer */
 static u32_t dhcpv4_send_request(struct net_if *iface)
 {
+	const struct in_addr *src = NULL;
```
Comment (tbursztyka) [inline subsys/net/ip/dhcpv4.c]: move it below ciaddr

**Review by tbursztyka — CHANGES_REQUESTED** [review]
small issues

File: subsys/net/ip/dhcpv4.c [hunk 5]
```diff
@@ -150,15 +150,22 @@ static inline bool dhcpv4_add_sname(struct net_pkt *pkt)
 /* Create DHCPv4 message and add options as per message type */
 static struct net_pkt *dhcpv4_create_message(struct net_if *iface, u8_t type,
 					     const struct in_addr *ciaddr,
+					     const struct in_addr *src,
 					     const struct in_addr *server_addr,
 					     bool server_id, bool requested_ip)
 {
 	NET_PKT_DATA_ACCESS_DEFINE(dhcp_access, struct dhcp_msg);
-	const struct in_addr src = INADDR_ANY_INIT;
+	const struct in_addr any = INADDR_ANY_INIT;
```
Comment (jukkar) [inline subsys/net/ip/dhcpv4.c]: This "any" variable can be removed as there exists a function `net_ipv4_unspecified_address()` that will return the any address to the caller.

File: subsys/net/ip/dhcpv4.c [hunk 6]
```diff
@@ -150,15 +150,22 @@ static inline bool dhcpv4_add_sname(struct net_pkt *pkt)
...
+	const struct in_addr *addr;
 	size_t size = DHCPV4_MESSAGE_SIZE;
 	struct net_pkt *pkt;
 	struct dhcp_msg *msg;
 
+	if (src == NULL)
+		addr = &any;
```
Comment (jukkar) [inline subsys/net/ip/dhcpv4.c]: so just do `addr = net_ipv4_unspecified_address()` here.

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/ip/dhcpv4.c (hunk 1 above)
Comment (pfl) [reply subsys/net/ip/dhcpv4.c]: Oh yes. Someone needs to file a bug against checkpatch as this is not detected automatically...!

File: subsys/net/ip/dhcpv4.c (hunk 2 above)
Comment (pfl) [reply subsys/net/ip/dhcpv4.c]: Fixed

File: subsys/net/ip/dhcpv4.c (hunk 4 above)
Comment (pfl) [reply subsys/net/ip/dhcpv4.c]: Yes

File: subsys/net/ip/dhcpv4.c (hunk 3 above)
Comment (pfl) [reply subsys/net/ip/dhcpv4.c]: Yes

File: subsys/net/ip/dhcpv4.c (hunk 5 above)
Comment (pfl) [reply subsys/net/ip/dhcpv4.c]: Done, that's a better one to use.

File: subsys/net/ip/dhcpv4.c (hunk 6 above)
Comment (pfl) [reply subsys/net/ip/dhcpv4.c]: Yes.

**Review by tbursztyka — APPROVED** [review]

**Review by rveerama1 — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Comment by pfl** [comment]
To clarify: Linux user space never sees the DHCP Request messages sent with source address 0.0.0.0. It will eventually work, as the Zephyr DHCP client restarts the address acquisition procedure with a DHCP Discover, but depending on the situation and DHCP servers involved, Zephyr might loose its address in between or get a new one assigned due to restarting of the DHCP procedure.

**Comment by pfl** [comment]
It would probably be a good idea to get this merged, I'd guess users would need this fix in the LTS anyway later.

**Review by pfalcon — APPROVED** [review]
This looks like a clean fix, and probably would (severely) affect production, long-running devices. @pfl, Please post a bug describing that case, and add "Fixes: " header to commit, to satisfy conditions for its inclusion in the release.

**Pull request merged**